### ISSUE TITLE
FAQ-D: going from prototype or basic model to full model

### ISSUE THREAD

**Issue by josef-pkt** [statsmodels/statsmodels#2705]
related what's required #1970  (doesn't provide much information yet)
(I thought I wrote already a list of "What's in a model and results?" but don't find it.)

Let's assume we want to build a new maximum likelihood model. For quick prototyping we can subclass `GenericLikelihoodModel` which tries to fill in some of the missing elements in a generic way.
Full MLE model subclass `LikelihoodModel` and `LikelihoodModelResuts` which requires some additional model specific parts.

The Base Model

The minimal part of any model is the loglikelihood function and a predict method, if prediction applies. LikelihoodModel provides data and formula handling and optimization,  the LikelihoodModelResults class provides inference based on Wald tests and a few extra methods.
(TODO: what else do we have generic? summary?)

These inherited methods are still mainly designed for the basic `(endog, exog, keywords)` models, and will need adjustments in models that follow a different pattern.
(TODO: explain what to do in two part models, and figure out which support is missing)

The basic model should then provide wald parameter inference, standard errors, tvalues and pvalues, and prediction.
general Wald tests, `t_test`, `wald_test`, are inherited but some things need to be available, e.g. t_test raises if `df_resid` is not defined.

The full model
------------------

- wrapper class for results: The wrapper class for a results class provides the pandas interface in the returns and attributes. For example the parameters `params` are returned as a pandas.Series if the original data was in a pandas DataFrame or Series.

(more later)

**Comment by datnamer** [comment]
This is great. Perhaps a new class like "  ScikitLearnRegressionModel can subclass both the needed classes and the original sklearn model and get fit info, walt test etc

**Comment by josef-pkt** [comment]
@datnamer  I'm not very familiar with the details of scikit-learn, so based on just coarse information.
It will not be possible to inherit from both since the structure is quite different. 

Somebody wrote a scikit-learn compatible wrapper once for statsmodels GLM to add to a pipeline or automatic cross-validation. (I don't know what the status of those approaches is.)

In the other direction to add inference to scikit-learn models: 
One problem for automatically adding inference is that scikit-learn doesn't calculate the necessary information, AFAIK, for example the Hessian and Jacobian of the loglikelihood function, or whatever equivalent we need. It would be possible to add those for a model selected and estimated by scikit-learn. if we can replicate the final model, otherwise I only see bootstrap parameter covariance as generic solution.
(Inference based on the selected model ignores the selection process, which means inference may not be accurate, except under some assumptions. We are still planning to do it this way in our own penalization until inference for variable selection can catch up, that's very recent work even in statistics and econometrics theory.)

No that I think about it, GenericLikelihoodModel does have an option for bootstrap parameter covariance, but I added it a long time ago as experimental feature. So I guess something like GenericLikelihoodModel could be rewritten to handle a scikit-learn model. Formula, datahandling and bootstrap cov_params based inference could be inherited. 
Our usual limitation, like using dense arrays in memory still apply unless or until we can overwrite or swap those parts.

**Comment by josef-pkt** [comment]
some notes on discrete_models

Initially written as part of an answer to the stackoverflow question What is mlefit in DiscreteResults or `LogitResults.__init__`. 


Browsing the code, it's slightly messy:

The `fit` method in the discrete models obtains `mlefit` from the super class and all attributes are attached to the results instance, i.e. `DiscreteResults.__init__` contains a generic

`self.__dict__.update(mlefit.__dict__)`

The base class that contains the `fit` method with the nonlinear optimizer is `statsmodels.base.LikelihoodModel` which at the end creates a basic `statsmodels.base.LikelihoodModelResults` instance. This in turn requires `params` and `normalized_cov_params` for the Regression results

`__init__(self, model, params, normalized_cov_params=None, scale=1., **kwargs)`

Note: `normalized_cov_params` is the, possibly unscaled, inverse Hessian for discrete models or maximum likelihoodmodels in general. In discrete_models it is already scaled correctly, so scale=1 which is the default.

So the main answer to the question is:

Based on this, `mlefit` needs to be an instance of a class that has at least `params` and `normalized_cov_params` as attributes or accessible through the `__dict__` argument. There might be other model specific attributes that are attached through the `kwargs` in the generic framework. Some keywords in the kwargs might be required by specific models but mainly for extra features.
(Sometimes I need to go by trial and error to see if I missed some attributes or keywords that are required for a results method that I want to use.)

This part is not well defined, because there are model type specific features that are not generically available and not required for all models, and because the flexibility is needed to add new model categories. Another reason for making things optional with the `kwargs` is the "lazy" policy that attributes are only required if they are used, i.e. if only `predict` is needed then `normalized_cov_params` could be `None`. (Or phrased more negatively: organic growth and no up-front design.)

Nevertheless, this works well within the current class structure, e.g. the fit methods of individual models in discrete are mainly a super call plus choosing the matching results class.
